core/src/Dscp/Core/FairCV.hs

```diff
@@ -0,0 +1,157 @@
+module Dscp.Core.FairCV
+       (
+         -- * Tagged proofs
+         TaggedProof
+       , Unchecked
+       , Valid
+       , mkTaggedProof
+       , unTaggedProof
+
+       , validateTaggedProof
+       , mergeProofs
+
+         -- * Fair CV
+       , FairCV (..)
+       , validateFairCV
+       , singletonFCV
+       , mergeFairCVs
+       , addProof
+
+         -- * Fair CV check result
+       , FairCVCheckResult (..)
+       , fullyValid
+       ) where
+
+import qualified Data.Map.Merge.Strict as M
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as M
+import Fmt (build, mapF, (+|), (|+))
+
+import Dscp.Core.Foundation
+import Dscp.Crypto
+
+---------------------------------------------------------------------------
+-- Tagged Merkle proofs
+---------------------------------------------------------------------------
+
+-- | Type tag for proofs/CVs not checked yet.
+data Unchecked
+
+-- | Type tag for validated proofs/CVs.
+data Valid
+
+-- | 'MerkleProof' tagged by validation tag. When using this module,
+-- it is impossible to obtain a value of 'TaggedProof Valid a' without
+-- doing a validity check.
+newtype TaggedProof v a = TaggedProof
+    { unTaggedProof :: MerkleProof a
+    } deriving (Show, Eq, Generic, Buildable)
+
+-- | Safe constructor for 'TaggedProof', which can yield only 'Unckecked'
+-- proof.
+mkTaggedProof :: MerkleProof a -> TaggedProof Unchecked a
+mkTaggedProof = TaggedProof
+
+-- | If 'Unchecked' proof is valid, make it 'Valid', otherwise return an
+-- error.
+validateTaggedProof
+    :: HasHash a
+    => TaggedProof Unchecked a
+    -> Either Text (TaggedProof Valid a)
+validateTaggedProof (TaggedProof unchecked) =
+    if isJust (computeMerkleRoot unchecked)
+    then Right $ TaggedProof unchecked
+    else Left "Proof validation has failed"
+
+-- | Merges two already validated Merkle proofs, checking that all common
+-- inner nodes match.
+mergeProofs
+    :: TaggedProof Valid a
+    -> TaggedProof Valid a
+    -> Either Text (TaggedProof Valid a)
+mergeProofs (TaggedProof a) (TaggedProof b)
+    | pnSig a /= pnSig b = Left "Node signatures mismatch"
+    | otherwise = TaggedProof <$> merge a b
+  where
+    merge (ProofBranch s l r) (ProofBranch _ l' r') =
+        let mergeUntag t t' =
+                unTaggedProof <$>
+                mergeProofs (TaggedProof t) (TaggedProof t')
+        in ProofBranch s <$> mergeUntag l l' <*> mergeUntag r r'
```

Comment:
By the way, you do not check that the signatures on the left and on the right sides match.
I suppose it means that merging two validated proofs can produce invalid proof if this function is used inaccurately.
Would be nice to specify more concrete preconditions in haddock.

Reply:
@Martoon-00 what do you mean? signature match is checked for _every_ pair of nodes, see line 73.

Reply:
Oh right, missed it this time.